File: Sources/DPIAHelpers/dpia_test_helpers.swift
// Stack 5 — Swift/URLSession DPIA test helpers
// Mirrors dpia_test_helpers.mjs for use in XCTest-based DPIA suites.

import Foundation

// MARK: - Environment

public func env(_ key: String, _ fallback: String = "") -> String {
    ProcessInfo.processInfo.environment[key] ?? fallback
}

// MARK: - Auth

public struct CognitoTokenFetcher {
    let region: String
    let clientId: String
    let password: String
    public init(region: String, clientId: String, password: String) {
        self.region = region; self.clientId = clientId; self.password = password
    }

    public func fetchToken(username: String) async throws -> String {
        let url = URL(string: "https://cognito-idp.\(region).amazonaws.com/")!
        var req = URLRequest(url: url)
        req.httpMethod = "POST"
        req.setValue("application/x-amz-json-1.1", forHTTPHeaderField: "Content-Type")
        req.setValue("AWSCognitoIdentityProviderService.InitiateAuth", forHTTPHeaderField: "X-Amz-Target")
        let body: [String: Any] = [
            "AuthFlow": "USER_PASSWORD_AUTH",
            "ClientId": clientId,
            "AuthParameters": ["USERNAME": username, "PASSWORD": password]
        ]
        req.httpBody = try JSONSerialization.data(withJSONObject: body)
        let (data, _) = try await URLSession.shared.data(for: req)
        let json = try JSONSerialization.jsonObject(with: data) as? [String: Any]
        let authResult = json?["AuthenticationResult"] as? [String: Any]
        guard let token = authResult?["IdToken"] as? String ?? authResult?["AccessToken"] as? String else {
            throw NSError(domain: "DPIAAuth", code: 1,
                userInfo: [NSLocalizedDescriptionKey: "Cognito failed: \(String(data: data, encoding: .utf8) ?? "")"])
        }
        return token
    }
}

// MARK: - HTTP client

public struct DPIAResponse {
    public let status: Int
    public let body: [String: Any]
    public init(status: Int, body: [String: Any]) { self.status = status; self.body = body }
}

public func dpiaRequest(
    method: String,
    url: String,
    body: [String: Any]?,
    token: String?,
    devUid: String?
) async throws -> DPIAResponse {
    guard let requestURL = URL(string: url) else {
        throw NSError(domain: "DPIARequest", code: 0,
            userInfo: [NSLocalizedDescriptionKey: "Invalid URL: \(url)"])
    }
    var req = URLRequest(url: requestURL)
    req.httpMethod = method
    req.setValue("application/json", forHTTPHeaderField: "Content-Type")
    if let token = token {
        req.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
    } else if let devUid = devUid {
        req.setValue(devUid, forHTTPHeaderField: "x-dev-uid")
    }
    if let body = body {
        req.httpBody = try JSONSerialization.data(withJSONObject: body)
    }
    let (data, response) = try await URLSession.shared.data(for: req)
    let httpResponse = response as! HTTPURLResponse
    let json = (try? JSONSerialization.jsonObject(with: data) as? [String: Any]) ?? [:]
    return DPIAResponse(status: httpResponse.statusCode, body: json)
}

// MARK: - Results model

public struct DPIAAssertion: Codable {
    public var label: String
    public var claim: String
    public var result: String   // "pass" | "fail" | "pending"
    public var ms: Int

    public init(label: String, claim: String) {
        self.label = label
        self.claim = claim
        self.result = "pending"
        self.ms = 0
    }
}

public struct DPIASection: Codable {
    public var name: String
    public var assertions: [DPIAAssertion]
    public init(name: String) { self.name = name; self.assertions = [] }
}

public struct DPIAResults: Codable {
    public var stack: String
    public var run: String
    public var githubRunId: String?
    public var timestamp: String
    public var auth: String
    public var ledger: String
    public var sections: [DPIASection]
    public init(stack: String, run: String, githubRunId: String?, timestamp: String, auth: String, ledger: String, sections: [DPIASection] = []) {
        self.stack = stack; self.run = run; self.githubRunId = githubRunId
        self.timestamp = timestamp; self.auth = auth; self.ledger = ledger; self.sections = sections
    }
}

// MARK: - Section builder

private enum DPIAResultStore {
    static var sections: [DPIASection] = []

    static func sectionIndex(named name: String) -> Int {
        if let index = sections.firstIndex(where: { $0.name == name }) {
            return index
        }
        sections.append(DPIASection(name: name))
        return sections.count - 1
    }

    static func append(_ assertion: DPIAAssertion, toSectionAt index: Int) {
        guard sections.indices.contains(index) else { return }
        sections[index].assertions.append(assertion)
    }
}

public class DPIASectionBuilder {
    public typealias ResultsMutator = (inout DPIAResults) -> Void
    private let mutate: ResultsMutator

    private var sectionIndex: Int = -1
    private var resultsRef: UnsafeMutablePointer<DPIAResults>

    public init(resultsPointer: UnsafeMutablePointer<DPIAResults>, sectionName: String) {
        self.resultsRef = resultsPointer
        self.mutate = { _ in }
        let sec = DPIASection(name: sectionName)
        resultsPointer.pointee.sections.append(sec)
        sectionIndex = resultsPointer.pointee.sections.count - 1
        _ = DPIAResultStore.sectionIndex(named: sectionName)
    }

    public func record(label: String, claim: String, start: Date, passed: Bool) {
        let ms = Int(Date().timeIntervalSince(start) * 1000)
        let assertion = DPIAAssertion(label: label, claim: claim)
        var updated = assertion
        updated.result = passed ? "pass" : "fail"
        updated.ms = ms
        resultsRef.pointee.sections[sectionIndex].assertions.append(updated)
        DPIAResultStore.append(updated, toSectionAt: DPIAResultStore.sectionIndex(named: resultsRef.pointee.sections[sectionIndex].name))
    }

    /// Overload accepting a millisecond timestamp (e.g. `Int64(Date().timeIntervalSince1970 * 1000)`).
    public func record(label: String, claim: String, startMs: Int64, passed: Bool) {
        let nowMs = Int64(Date().timeIntervalSince1970 * 1000)
        let assertion = DPIAAssertion(label: label, claim: claim)
        var updated = assertion
        updated.result = passed ? "pass" : "fail"
        updated.ms = Int(nowMs - startMs)
        resultsRef.pointee.sections[sectionIndex].assertions.append(updated)
        DPIAResultStore.append(updated, toSectionAt: DPIAResultStore.sectionIndex(named: resultsRef.pointee.sections[sectionIndex].name))
    }
}

// MARK: - Write results

public func writeResults(results: DPIAResults, outDir: String, stackNum: String) {
    var output = results
    if output.sections.allSatisfy({ $0.assertions.isEmpty }) && !DPIAResultStore.sections.isEmpty {
        output.sections = DPIAResultStore.sections
    }
    let fm = FileManager.default
    try? fm.createDirectory(atPath: outDir, withIntermediateDirectories: true)
    let outPath = (outDir as NSString).appendingPathComponent("unit-results-stack\(stackNum).json")
    let encoder = JSONEncoder()
    encoder.outputFormatting = [.prettyPrinted, .sortedKeys]
    if let data = try? encoder.encode(output) {
        try? data.write(to: URL(fileURLWithPath: outPath))
    }
    let total = output.sections.reduce(0) { $0 + $1.assertions.count }
    let passed = output.sections.reduce(0) { $0 + $1.assertions.filter { $0.result == "pass" }.count }
    print("\n  Unit results: \(passed)/\(total) assertions passed")
    print("  Written: \(outPath)")
    for sec in output.sections {
        for a in sec.assertions {
            let icon = a.result == "pass" ? "✓" : a.result == "pending" ? "○" : "✗"
            print("    \(icon) [\(sec.name)] \(a.label)")
        }
    }
}

File: Sources/DPIAHelpers/dpia_test_helpers.kt
// Stack 6 — Kotlin/OkHttp DPIA test helpers
// Mirrors dpia_test_helpers.mjs: four plain functions, no class wrappers.
// Each test file defines its own req() inline; only shared utilities live here.

package dpia

import kotlinx.serialization.Serializable
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.*
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import java.io.File
import java.security.MessageDigest

// ── Environment ───────────────────────────────────────────────────────────────

fun env(key: String, fallback: String = ""): String = System.getenv(key) ?: fallback

// ── SHA-256 ───────────────────────────────────────────────────────────────────

fun sha256Hex(input: String): String =
    MessageDigest.getInstance("SHA-256").digest(input.toByteArray())
        .joinToString("") { "%02x".format(it) }

// ── Cognito ───────────────────────────────────────────────────────────────────

class CognitoTokenFetcher(
    private val region: String,
    private val clientId: String,
    private val password: String
) {
    private val client = OkHttpClient()

    fun fetchToken(username: String): String {
        val body = """{"AuthFlow":"USER_PASSWORD_AUTH","ClientId":"$clientId","AuthParameters":{"USERNAME":"$username","PASSWORD":"$password"}}"""
        val req = Request.Builder()
            .url("https://cognito-idp.$region.amazonaws.com/")
            .post(body.toRequestBody("application/x-amz-json-1.1".toMediaType()))
            .addHeader("X-Amz-Target", "AWSCognitoIdentityProviderService.InitiateAuth")
            .build()
        val resp = client.newCall(req).execute()
        val json = Json.parseToJsonElement(resp.body!!.string()).jsonObject
        val auth = json["AuthenticationResult"]?.jsonObject
            ?: error("Cognito failed: ${resp.body}")
        return ((auth["IdToken"] ?: auth["AccessToken"]) as JsonPrimitive).content
    }
}

// ── HTTP response ─────────────────────────────────────────────────────────────
// Plain data class — mirrors Stack 4's { status, body } object.

data class Res(val status: Int, val body: Map<String, Any?>)

// ── HTTP execution (shared by each test's inline req()) ───────────────────────

fun makeRequest(
    client: OkHttpClient,
    method: String,
    url: String,
    body: Map<String, Any?>?,
    headers: Map<String, String>
): Res {
    val rb = Request.Builder().url(url).addHeader("Content-Type", "application/json")
    headers.forEach { (k, v) -> rb.addHeader(k, v) }
    val okBody = body?.let { jsonOf(it).toRequestBody("application/json".toMediaType()) }
    val req = when (method.uppercase()) {
        "GET"    -> rb.get()
        "DELETE" -> if (okBody != null) rb.delete(okBody) else rb.delete()
        "POST"   -> rb.post(okBody ?: "{}".toRequestBody("application/json".toMediaType()))
        "PUT"    -> rb.put(okBody ?: "{}".toRequestBody("application/json".toMediaType()))
        else     -> rb.method(method, okBody)
    }.build()
    val resp = client.newCall(req).execute()
    return Res(resp.code, parseBody(resp.body?.string() ?: "{}"))
}

// ── JSON utilities ────────────────────────────────────────────────────────────

fun jsonOf(v: Any?): String = when (v) {
    null         -> "null"
    is String    -> "\"${v.replace("\\", "\\\\").replace("\"", "\\\"")}\""
    is Number, is Boolean -> v.toString()
    is Map<*, *> -> "{${v.entries.joinToString(",") { (k, w) -> "\"$k\":${jsonOf(w)}" }}}"
    is List<*>   -> "[${v.joinToString(",") { jsonOf(it) }}]"
    else         -> "\"$v\""
}

fun parseBody(json: String): Map<String, Any?> = try {
    (Json { ignoreUnknownKeys = true }.parseToJsonElement(json) as? JsonObject)
        ?.let { it.entries.associate { (k, v) -> k to v.toPlain() } } ?: emptyMap()
} catch (_: Exception) { emptyMap() }

fun JsonElement.toPlain(): Any? = when (this) {
    is JsonNull      -> null
    is JsonPrimitive -> when {
        isString              -> content
        booleanOrNull != null -> booleanOrNull
        intOrNull != null     -> intOrNull
        longOrNull != null    -> longOrNull
        doubleOrNull != null  -> doubleOrNull
        else                  -> content
    }
    is JsonObject -> entries.associate { (k, v) -> k to v.toPlain() }
    is JsonArray  -> map { it.toPlain() }
}

fun Map<String, Any?>.str(key: String): String? = when (val v = this[key]) {
    null      -> null
    is String -> v
    else      -> v.toString()
}

fun Map<String, Any?>.ledger(): Map<*, *> =
    this["ledger"] as? Map<*, *> ?: this

fun Map<String, Any?>.ledgerInt(key: String): Int =
    when (val v = ledger()[key]) {
        is Number -> v.toInt()
        is String -> v.toIntOrNull() ?: 0
        else      -> 0
    }

fun Map<String, Any?>.ledgerBool(key: String): Boolean? =
    when (val v = ledger()[key]) {
        is Boolean -> v
        is String  -> v.toBooleanStrictOrNull()
        else       -> null
    }

// ── Results ───────────────────────────────────────────────────────────────────
// Mirrors Stack 4's { stack, run, sections: [{ name, assertions: [] }] }

@Serializable
data class DPIAAssertion(
    val label: String,
    val claim: String,
    var result: String = "pending",
    var ms: Long = 0
)

@Serializable
data class DPIASection(
    val name: String,
    val assertions: MutableList<DPIAAssertion> = mutableListOf()
)

@Serializable
data class DPIAResults(
    val stack: String,
    val run: String,
    val githubRunId: String? = null,
    val timestamp: String,
    val auth: String,
    val ledger: String,
    val sections: MutableList<DPIASection> = mutableListOf()
)

private object DPIAResultStore {
    val sections: MutableList<DPIASection> = mutableListOf()

    fun section(name: String): DPIASection =
        sections.find { it.name == name }
            ?: DPIASection(name = name).also { sections.add(it) }
}

fun record(
    results: DPIAResults,
    section: String,
    label: String,
    claim: String,
    startMs: Long,
    passed: Boolean
) {
    val sec = results.sections.find { it.name == section }
        ?: DPIASection(name = section).also { results.sections.add(it) }
    val assertion = DPIAAssertion(
        label  = label,
        claim  = claim,
        result = if (passed) "pass" else "fail",
        ms     = System.currentTimeMillis() - startMs
    )
    sec.assertions.add(assertion)
    DPIAResultStore.section(section).assertions.add(assertion)
}

fun writeResults(results: DPIAResults, outDir: String, stackNum: String) {
    val output = if (results.sections.all { it.assertions.isEmpty() } && DPIAResultStore.sections.isNotEmpty()) {
        results.copy(sections = DPIAResultStore.sections)
    } else {
        results
    }
    val dir = File(outDir); dir.mkdirs()
    val outFile = File(dir, "unit-results-stack$stackNum.json")
    outFile.writeText(Json { prettyPrint = true }.encodeToString(output))
    val total  = output.sections.sumOf { it.assertions.size }
    val passed = output.sections.sumOf { s -> s.assertions.count { it.result == "pass" } }
    println("\n  Unit results: $passed/$total assertions passed")
    println("  Written: ${outFile.absolutePath}")
    for (sec in output.sections) for (a in sec.assertions) {
        println("    ${if (a.result == "pass") "✓" else "✗"} [${sec.name}] ${a.label}")
    }
}
